Annotate RemoteAudioSessionProxy message endpoints with feature flag

szewai opened the pull request:
#### 76bbaa0950c935aea5d06b58f5e6500051ddbb22
<pre>
Annotate RemoteAudioSessionProxy message endpoints with feature flag
<a href="https://bugs.webkit.org/show_bug.cgi?id=280238">https://bugs.webkit.org/show_bug.cgi?id=280238</a>
<a href="https://rdar.apple.com/136540877">rdar://136540877</a>

Reviewed by Eric Carlson.

Add a feature flag for guarding media playback functionalities and annotate RemoteAudioSessionProxy messages with it.
This means when media playback is disabled, web process should not be sending RemoteAudioSessionProxy messages to GPU
process.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/platform/audio/AudioSession.cpp:
(WebCore::AudioSession::setEnabled):
* Source/WebCore/platform/audio/AudioSession.h:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.cpp:
(WebKit::RemoteAudioSessionProxy::sharedPreferencesForWebProcess const):
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.h:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.messages.in:
* Source/WebKit/Scripts/PreferencesTemplates/WebPageUpdatePreferences.cpp.erb:
* Source/WebKit/Scripts/webkit/messages.py:
(generate_enabled_by_for_receiver):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByAndConjunctionMessageReceiver.cpp:
(WebKit::TestWithEnabledByAndConjunction::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessageReceiver.cpp:
(WebKit::TestWithEnabledBy::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByOrConjunctionMessageReceiver.cpp:
(WebKit::TestWithEnabledByOrConjunction::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledIfMessageReceiver.cpp:
(WebKit::TestWithEnabledIf::didReceiveMessage):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.cpp:
(WebKit::RemoteAudioSession::ensureConnection):
(WebKit::RemoteAudioSession::configuration):
(WebKit::RemoteAudioSession::setCategory):
(WebKit::RemoteAudioSession::setPreferredBufferSize):
(WebKit::RemoteAudioSession::setIsPlayingToBluetoothOverride):
(WebKit::RemoteAudioSession::beginAudioSessionInterruption):
(WebKit::RemoteAudioSession::endAudioSessionInterruption):
(WebKit::RemoteAudioSession::beginInterruptionForTesting):
(WebKit::RemoteAudioSession::endInterruptionForTesting):
(WebKit::RemoteAudioSession::setSceneIdentifier):
(WebKit::RemoteAudioSession::setSoundStageSize):
(WebKit::RemoteAudioSession::setEnabled):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updatePreferences):

Canonical link: <a href="https://commits.webkit.org/284331@main">https://commits.webkit.org/284331@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/30eb0657ff140c8c693702a81e50f061013d0c6f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69068 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48468 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21740 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73149 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20226 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71185 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56269 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20075 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54986 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13428 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72134 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44244 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59631 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35463 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/68577 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40910 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17061 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18600 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/62184 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62858 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17406 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74860 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/68314 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13050 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16651 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62633 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13088 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59714 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62532 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15326 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10523 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4135 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/90095 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44272 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15975 "Found 38 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Number/property_and_index_of_number.js.default, ChakraCore.yaml/ChakraCore/test/fieldopts/fixedfieldmonocheck.js.default, ChakraCore.yaml/ChakraCore/test/typedarray/bug2230916.js.default, stress/arguments-non-configurable.js.bytecode-cache, stress/async-await-basic.js.dfg-eager, stress/async-iteration-for-await-of.js.dfg-eager, stress/big-int-bitwise-or-type-error.js.dfg-eager, stress/big-int-bitwise-or-type-error.js.mini-mode, stress/big-int-exp-negative-exponent.js.bytecode-cache, stress/big-int-exp-negative-exponent.js.dfg-eager ... (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45345 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46541 "Built successfully") | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45087 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.cpp:
```diff
@@ -196,6 +196,12 @@ void RemoteAudioSessionProxy::triggerEndInterruptionForTesting()
 
+const SharedPreferencesForWebProcess& RemoteAudioSessionProxy::sharedPreferencesForWebProcess() const
+{
+    RefPtr gpuConnectionToWebProcess = m_gpuConnection.get();
+    return gpuConnectionToWebProcess->sharedPreferencesForWebProcess();
```
There are two problems here:
* m_gpuConnection is a ThreadSafeWeakPtr<> so it can be null and the dereference will crash
* the GPU process connection may be destroyed after you return a Ref to the SharedPreferencesForWebProcess object it owns

szewai replied:
> m_gpuConnection is a ThreadSafeWeakPtr<> so it can be null and the dereference will crash

The crash is expected if connection is null because the function is currently only invoked by message handlers. If the connection doesn't exist (or is already gone), then the message handlers should not be invoked. 

I think we could this more robust by have it returning a std::optional<SharedPreferencesForWebProcess> and make message handlers null check it, but that will require all the other places that has this function to change; I will do it in a followup.

> the GPU process connection may be destroyed after you return a Ref to the SharedPreferencesForWebProcess object it owns

This is good point; I will make the function returns a copy instead.

eric-carlson replied:
> The crash is expected if connection is null because the function is currently only invoked by message handlers. If the connection doesn't exist (or is already gone), then the message handlers should not be invoked.
A `RELEASE_ASSERT` would be preferable to a crash because of a null deref

cdumez replied:
> A RELEASE_ASSERT would be preferable to a crash because of a null deref

News to me. A null deref is a safe & very actionable crash. The RELEASE_ASSERT() would add extra cost for no great reason in my opinion. As a matter of fact, reviewers will often point out that some assertions are useless in this extra case.

szewai replied:
Yeah I've added RELEASE_ASSERT before and Andy said it's unnecessary since they are both crash: https://github.com/WebKit/WebKit/pull/33703#discussion_r1763759712.
Do you think I should a release assert here? It should be temporary anyways, I will make the function return optional in the following change.

eric-carlson replied:
>Do you think I should a release assert here?

No, I would trust @cdumez's opinion here over mine

eric-carlson commented on Source/WebKit/WebProcess/Storage/WebSharedWorkerContextManagerConnection.cpp:
```diff
@@ -96,6 +96,11 @@ void WebSharedWorkerContextManagerConnection::updatePreferencesStore(const WebPr
+#if USE(AUDIO_SESSION)
+    if (m_preferencesStore->getBoolValueForKey(WebPreferencesKey::mediaPlaybackEnabledKey()))
+        AudioSession::sharedSession().setEnabled(true);
+#endif
```
I don't think it makes sense to enable/disable the audio session here. Is there a reason it isn't done in `WebPage::updatePreferencesGenerated`?

szewai replied:
Not really, I could add it to `WebPage::updatePreferencesGenerated`.

szewai committed:
Annotate RemoteAudioSessionProxy message endpoints with feature flag
https://bugs.webkit.org/show_bug.cgi?id=280238
rdar://136540877

Reviewed by Eric Carlson.

Add a feature flag for guarding media playback functionalities and annotate RemoteAudioSessionProxy messages with it.
This means when media playback is disabled, web process should not be sending RemoteAudioSessionProxy messages to GPU
process.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/platform/audio/AudioSession.cpp:
(WebCore::AudioSession::setEnabled):
* Source/WebCore/platform/audio/AudioSession.h:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.cpp:
(WebKit::RemoteAudioSessionProxy::sharedPreferencesForWebProcess const):
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.h:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.messages.in:
* Source/WebKit/Scripts/PreferencesTemplates/WebPageUpdatePreferences.cpp.erb:
* Source/WebKit/Scripts/webkit/messages.py:
(generate_enabled_by_for_receiver):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByAndConjunctionMessageReceiver.cpp:
(WebKit::TestWithEnabledByAndConjunction::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessageReceiver.cpp:
(WebKit::TestWithEnabledBy::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByOrConjunctionMessageReceiver.cpp:
(WebKit::TestWithEnabledByOrConjunction::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledIfMessageReceiver.cpp:
(WebKit::TestWithEnabledIf::didReceiveMessage):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.cpp:
(WebKit::RemoteAudioSession::ensureConnection):
(WebKit::RemoteAudioSession::configuration):
(WebKit::RemoteAudioSession::setCategory):
(WebKit::RemoteAudioSession::setPreferredBufferSize):
(WebKit::RemoteAudioSession::setIsPlayingToBluetoothOverride):
(WebKit::RemoteAudioSession::beginAudioSessionInterruption):
(WebKit::RemoteAudioSession::endAudioSessionInterruption):
(WebKit::RemoteAudioSession::beginInterruptionForTesting):
(WebKit::RemoteAudioSession::endInterruptionForTesting):
(WebKit::RemoteAudioSession::setSceneIdentifier):
(WebKit::RemoteAudioSession::setSoundStageSize):
(WebKit::RemoteAudioSession::setEnabled):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updatePreferences):

Canonical link: https://commits.webkit.org/284331@main
